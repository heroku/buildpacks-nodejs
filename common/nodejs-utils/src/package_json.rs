use crate::vrs::{Requirement, Version};
use serde::Deserialize;
use std::collections::HashMap;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use thiserror::Error;

#[derive(Deserialize, Debug, Clone)]
pub struct PackageJson {
    pub name: String,
    pub version: Option<Version>,
    pub engines: Option<Engines>,
    pub scripts: Option<Scripts>,
    pub main: Option<String>,
    pub dependencies: Option<HashMap<String, String>>,
    #[serde(rename = "devDependencies")]
    pub dev_dependencies: Option<HashMap<String, String>>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct Engines {
    pub node: Option<Requirement>,
    pub yarn: Option<Requirement>,
    pub npm: Option<Requirement>,
}

#[derive(Deserialize, Debug, Clone)]
#[serde(rename_all = "kebab-case")]
pub struct Scripts {
    pub start: Option<String>,
    pub build: Option<String>,
    pub heroku_prebuild: Option<String>,
    pub heroku_build: Option<String>,
    pub heroku_postbuild: Option<String>,
}

#[derive(Error, Debug)]
pub enum PackageJsonError {
    #[error("Could not read package.json. {0}")]
    AccessError(std::io::Error),
    #[error("Could not parse package.json. {0}")]
    ParseError(serde_json::Error),
}

impl PackageJson {
    /// Reads a package.json from a path into a `PackageJson` struct.
    ///
    /// # Errors
    ///
    /// * Invalid/malformed JSON
    /// * Path does not exist or is unreadable
    /// * Version strings are invalid/malformed
    pub fn read<P: AsRef<Path>>(path: P) -> Result<Self, PackageJsonError> {
        let file = File::open(path).map_err(PackageJsonError::AccessError)?;
        let rdr = BufReader::new(file);
        serde_json::from_reader(rdr).map_err(PackageJsonError::ParseError)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::Builder;

    #[test]
    fn read_valid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{\"name\": \"foo\",\"version\": \"0.0.0\"}}").unwrap();
        let pkg = PackageJson::read(f.path()).unwrap();
        assert_eq!(pkg.name, "foo");
        assert_eq!(pkg.version.unwrap().to_string(), "0.0.0");
    }

    #[test]
    fn read_valid_package_with_node_engine() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(
            f,
            "{{
            \"name\": \"foo\",
            \"version\": \"0.0.0\",
            \"engines\": {{
                \"node\": \"16.0.0\"
            }}
        }}"
        )
        .unwrap();
        let pkg = PackageJson::read(f.path()).unwrap();
        assert_eq!(&pkg.engines.unwrap().node.unwrap().to_string(), "16.0.0");
    }

    #[test]
    fn read_missing_package() {
        let res = PackageJson::read(Path::new("/over/there/package.json"));
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{}", err);
        assert!(err.contains("Could not read package.json"));
    }

    #[test]
    fn read_invalid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{").unwrap();
        let res = PackageJson::read(f.path());
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{}", err);
        assert!(err.contains("Could not parse package.json"));
    }
}
