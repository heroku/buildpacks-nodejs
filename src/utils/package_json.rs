use crate::utils::vrs::Requirement;
use serde::Deserialize;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use thiserror::Error;

#[derive(Deserialize, Debug, Default, Clone)]
pub(crate) struct PackageJson {
    pub(crate) engines: Option<Engines>,
    pub(crate) scripts: Option<Scripts>,
}

#[derive(Deserialize, Debug, Default, Clone)]
pub(crate) struct Engines {
    pub(crate) yarn: Option<Requirement>,
}

#[derive(Deserialize, Debug, Default, Clone)]
pub(crate) struct Scripts {
    pub(crate) start: Option<String>,
    pub(crate) build: Option<String>,
    #[serde(rename = "heroku-prebuild")]
    pub(crate) heroku_prebuild: Option<String>,
    #[serde(rename = "heroku-build")]
    pub(crate) heroku_build: Option<String>,
    #[serde(rename = "heroku-postbuild")]
    pub(crate) heroku_postbuild: Option<String>,
}

#[derive(Error, Debug)]
pub(crate) enum PackageJsonError {
    #[error("Could not read package.json. {0}")]
    AccessError(std::io::Error),
    #[error("Could not parse package.json. {0}")]
    ParseError(serde_json::Error),
}

impl PackageJson {
    /// Reads a package.json from a path into a `PackageJson` struct.
    ///
    /// # Errors
    ///
    /// * Invalid/malformed JSON
    /// * Path does not exist or is unreadable
    /// * Version strings are invalid/malformed
    pub(crate) fn read<P: AsRef<Path>>(path: P) -> Result<Self, PackageJsonError> {
        let file = File::open(path).map_err(PackageJsonError::AccessError)?;
        let rdr = BufReader::new(file);
        serde_json::from_reader(rdr).map_err(PackageJsonError::ParseError)
    }

    #[must_use]
    /// Fetches the build scripts from a `PackageJson` and returns them in priority order
    pub(crate) fn build_scripts(&self) -> Vec<String> {
        let mut scripts = vec![];
        if let Some(s) = &self.scripts {
            if s.heroku_prebuild.is_some() {
                scripts.push("heroku-prebuild".to_owned());
            }
            if s.heroku_build.is_some() {
                scripts.push("heroku-build".to_owned());
            } else if s.build.is_some() {
                scripts.push("build".to_owned());
            }
            if s.heroku_postbuild.is_some() {
                scripts.push("heroku-postbuild".to_owned());
            }
        }
        scripts
    }

    #[must_use]
    /// Determines if a given `PackageJson` has a start script defined
    pub(crate) fn has_start_script(&self) -> bool {
        self.scripts
            .as_ref()
            .is_some_and(|scripts| scripts.start.is_some())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::Builder;

    #[test]
    fn read_empty_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{ }}").unwrap();
        let pkg = PackageJson::read(f.path());
        assert!(pkg.is_ok());
    }

    #[test]
    fn read_valid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{\"name\": \"foo\",\"version\": \"0.0.0\"}}").unwrap();
        let pkg = PackageJson::read(f.path());
        assert!(pkg.is_ok());
    }

    #[test]
    fn read_missing_package() {
        let res = PackageJson::read(Path::new("/over/there/package.json"));
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{err}");
        assert!(err.contains("Could not read package.json"));
    }

    #[test]
    fn read_invalid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{").unwrap();
        let res = PackageJson::read(f.path());
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{err}");
        assert!(err.contains("Could not parse package.json"));
    }

    #[test]
    fn test_build_scripts_all() {
        let pkg_json = PackageJson {
            scripts: Some(Scripts {
                build: Some("echo 'build'".to_owned()),
                heroku_prebuild: Some("echo 'heroku-prebuild'".to_owned()),
                heroku_build: Some("echo 'build'".to_owned()),
                heroku_postbuild: Some("echo 'heroku-postbuild'".to_owned()),
                ..Scripts::default()
            }),
            ..PackageJson::default()
        };
        let build_scripts = pkg_json.build_scripts();

        assert_eq!("heroku-prebuild", build_scripts[0]);
        assert_eq!("heroku-build", build_scripts[1]);
        assert_eq!("heroku-postbuild", build_scripts[2]);
    }

    #[test]
    fn test_build_scripts_build_only() {
        let pkg_json = PackageJson {
            scripts: Some(Scripts {
                build: Some("echo 'build'".to_owned()),
                ..Scripts::default()
            }),
            ..PackageJson::default()
        };
        let build_scripts = pkg_json.build_scripts();

        assert_eq!("build", build_scripts[0]);
    }
}
