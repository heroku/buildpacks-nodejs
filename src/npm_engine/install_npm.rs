use super::main::NpmEngineBuildpackError;
use crate::utils::http::{ResponseExt, get};
use crate::utils::npm_registry::PackagePackument;
use crate::utils::vrs::Version;
use crate::{BuildpackBuildContext, BuildpackError, BuildpackResult};
use bullet_stream::global::print;
use fun_run::{CommandWithName, NamedOutput};
use libcnb::Env;
use libcnb::data::layer_name;
use libcnb::layer::{
    CachedLayerDefinition, EmptyLayerCause, InvalidMetadataAction, LayerState, RestoredLayerAction,
};
use libcnb::layer_env::Scope;
use libherokubuildpack::tar::decompress_tarball;
use serde::{Deserialize, Serialize};
use std::fs::File;
use std::path::{Path, PathBuf};
use std::process::Command;

pub(crate) fn install_npm(
    context: &BuildpackBuildContext,
    env: &Env,
    npm_packument: &PackagePackument,
    node_version: &Version,
) -> BuildpackResult<Env> {
    let npm_version = &npm_packument.version;

    let new_metadata = NpmEngineLayerMetadata {
        node_version: node_version.to_string(),
        npm_version: npm_version.to_string(),
        layer_version: LAYER_VERSION.to_string(),
        arch: context.target.arch.clone(),
        os: context.target.os.clone(),
    };

    let npm_engine_layer = context.cached_layer(
        layer_name!("npm_engine"),
        CachedLayerDefinition {
            build: true,
            launch: true,
            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
            restored_layer_action: &|old_metadata: &NpmEngineLayerMetadata, _| {
                if old_metadata == &new_metadata {
                    Ok((RestoredLayerAction::KeepLayer, vec![]))
                } else {
                    Ok((
                        RestoredLayerAction::DeleteLayer,
                        changed_metadata_fields(old_metadata, &new_metadata),
                    ))
                }
            },
        },
    )?;

    match npm_engine_layer.state {
        LayerState::Restored { .. } => {
            print::sub_bullet("Using cached npm");
        }
        LayerState::Empty { ref cause } => {
            npm_engine_layer.write_metadata(new_metadata)?;

            if let EmptyLayerCause::RestoredLayerAction { cause } = cause {
                print::sub_bullet(format!(
                    "Invalidating cached npm ({} changed)",
                    cause.join(", ")
                ));
            }

            let downloaded_package_path = npm_engine_layer.path().join("npm.tgz");
            let npm_cli_script = npm_engine_layer.path().join("package/bin/npm-cli.js");

            // this install process is generalized from the npm install script at:
            // https://www.npmjs.com/install.sh
            download_and_unpack_release(
                &npm_packument.dist.tarball,
                &downloaded_package_path,
                &npm_engine_layer.path(),
            )?;
            remove_existing_npm_installation(&npm_cli_script, env)?;
            install_npm_package(&npm_cli_script, &downloaded_package_path, env)?;
        }
    }

    Ok(npm_engine_layer.read_env()?.apply(Scope::Build, env))
}

fn download_and_unpack_release(
    download_from: &String,
    download_to: &Path,
    unpack_into: &Path,
) -> Result<(), NpmInstallError> {
    get(download_from)
        .call_sync()
        .and_then(|res| res.download_to_file_sync(download_to))
        .map_err(NpmInstallError::Download)
        .and_then(|()| {
            File::open(download_to)
                .map_err(|e| NpmInstallError::OpenTarball(download_to.to_path_buf(), e))
        })
        .and_then(|mut npm_tgz_file| {
            decompress_tarball(&mut npm_tgz_file, unpack_into)
                .map_err(|e| NpmInstallError::DecompressTarball(download_to.to_path_buf(), e))
        })?;
    Ok(())
}

fn remove_existing_npm_installation(
    npm_cli_script: &Path,
    env: &Env,
) -> Result<(), NpmInstallError> {
    print::sub_bullet("Removing npm bundled with Node.js");
    Command::new("node")
        .args([
            &npm_cli_script.to_string_lossy(),
            "rm",
            "npm",
            "-gf",
            "--loglevel=silent",
        ])
        .envs(env)
        .named_output()
        .map_err(NpmInstallError::RemoveExistingNpmInstall)
        .map(|_| ())
}

fn install_npm_package(
    npm_cli_script: &Path,
    package: &Path,
    env: &Env,
) -> Result<(), NpmInstallError> {
    print::sub_bullet("Installing requested npm");
    Command::new("node")
        .args([
            &npm_cli_script.to_string_lossy(),
            "install",
            "-gf",
            &package.to_string_lossy(),
        ])
        .envs(env)
        .named_output()
        .and_then(NamedOutput::nonzero_captured)
        .map_err(NpmInstallError::InstallNpm)
        .map(|_| ())
}

fn changed_metadata_fields(
    old: &NpmEngineLayerMetadata,
    new: &NpmEngineLayerMetadata,
) -> Vec<String> {
    let mut changed = vec![];
    if old.npm_version != new.npm_version {
        changed.push("npm version".to_string());
    }
    if old.node_version != new.node_version {
        changed.push("node version".to_string());
    }
    if old.layer_version != new.layer_version {
        changed.push("layer version".to_string());
    }
    if old.os != new.os {
        changed.push("operating system".to_string());
    }
    if old.arch != new.arch {
        changed.push("compute architecture".to_string());
    }
    changed.sort();
    changed
}

const LAYER_VERSION: &str = "1";

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub(crate) struct NpmEngineLayerMetadata {
    layer_version: String,
    npm_version: String,
    node_version: String,
    arch: String,
    os: String,
}

#[derive(Debug)]
pub(crate) enum NpmInstallError {
    Download(crate::utils::http::Error),
    OpenTarball(PathBuf, std::io::Error),
    DecompressTarball(PathBuf, std::io::Error),
    RemoveExistingNpmInstall(fun_run::CmdError),
    InstallNpm(fun_run::CmdError),
}

impl From<NpmInstallError> for NpmEngineBuildpackError {
    fn from(value: NpmInstallError) -> Self {
        NpmEngineBuildpackError::NpmInstall(value)
    }
}

impl From<NpmInstallError> for BuildpackError {
    fn from(value: NpmInstallError) -> Self {
        NpmEngineBuildpackError::from(value).into()
    }
}
