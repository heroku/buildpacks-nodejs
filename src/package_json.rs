use crate::utils::error_handling::{
    ErrorMessage, ErrorType, SuggestRetryBuild, SuggestSubmitIssue, error_message, file_value,
};
use crate::utils::vrs::{Requirement, Version, VersionError};
use bullet_stream::style;
use indoc::formatdoc;
use std::fmt::Display;
use std::path::{Path, PathBuf};
use std::str::FromStr;

pub(crate) struct PackageJson(serde_json::Value);

impl PackageJson {
    pub(crate) fn node_engine(&self) -> Option<Result<Requirement, VersionError>> {
        self.engines()
            .and_then(|engines| engines.get("node"))
            .and_then(|node| node.as_str())
            .map(Requirement::parse)
    }

    pub(crate) fn npm_engine(&self) -> Option<Requirement> {
        self.engines()
            .and_then(|engines| engines.get("npm"))
            .and_then(|node| node.as_str())
            .and_then(|node| Requirement::parse(node).ok())
    }

    fn engines(&self) -> Option<&serde_json::Value> {
        self.0.get("engines")
    }

    pub(crate) fn package_manager(
        &self,
    ) -> Option<Result<PackageManagerField, PackageManagerFieldError>> {
        self.0
            .get("packageManager")
            .and_then(|val| val.as_str())
            .map(PackageManagerField::from_str)
    }
}

impl TryFrom<PathBuf> for PackageJson {
    type Error = ErrorMessage;

    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
        let contents = std::fs::read_to_string(&value)
            .map_err(|e| package_json_read_error_message(&value, &e))?;
        let json = serde_json::from_str::<serde_json::Value>(&contents)
            .map_err(|e| package_json_parse_error_message(&value, &e))?;
        Ok(Self(json))
    }
}

fn package_json_read_error_message(path: &Path, error: &std::io::Error) -> ErrorMessage {
    let package_json = file_value(path);
    error_message()
        .error_type(ErrorType::UserFacing(
            SuggestRetryBuild::Yes,
            SuggestSubmitIssue::No,
        ))
        .header(format!("Error reading {package_json}"))
        .body(formatdoc! { "
            The Heroku Node.js buildpack reads from {package_json} to complete the build but \
            the file can't be read.

            Suggestions:
            - Ensure the file has read permissions.
        " })
        .debug_info(error.to_string())
        .create()
}

fn package_json_parse_error_message(path: &Path, error: &serde_json::Error) -> ErrorMessage {
    let package_json = file_value(path);
    let json_spec_url = style::url("https://www.json.org/");
    error_message()
        .error_type(ErrorType::UserFacing(
            SuggestRetryBuild::Yes,
            SuggestSubmitIssue::No,
        ))
        .header(format!("Error parsing {package_json}"))
        .body(formatdoc! { "
            The Heroku Node.js buildpack reads from {package_json} to complete the build but \
            the file isn't valid JSON.

            Suggestions:
            - Ensure the file follows the JSON format described at {json_spec_url}
        " })
        .debug_info(error.to_string())
        .create()
}

#[derive(Debug, PartialEq)]
pub(crate) struct PackageManagerField {
    pub(crate) name: PackageManagerFieldPackageManager,
    pub(crate) version: Version,
    integrity_check: Option<String>,
}

impl FromStr for PackageManagerField {
    type Err = PackageManagerFieldError;

    fn from_str(val: &str) -> Result<Self, Self::Err> {
        match val.split_once('@') {
            Some((name, remaining)) => {
                let name = PackageManagerFieldPackageManager::from_str(name).map_err(
                    |package_manager| PackageManagerFieldError::InvalidPackageManager {
                        package_manager,
                        field_value: val.to_owned(),
                    },
                )?;

                if let Some((version, integrity_check)) = remaining.split_once('+') {
                    let version = Version::from_str(version).map_err(|e| {
                        PackageManagerFieldError::Version {
                            field_value: val.to_owned(),
                            source: e,
                        }
                    })?;
                    Ok(Self {
                        name,
                        version,
                        integrity_check: Some(integrity_check.to_owned()),
                    })
                } else {
                    let version = Version::from_str(remaining).map_err(|e| {
                        PackageManagerFieldError::Version {
                            field_value: val.to_owned(),
                            source: e,
                        }
                    })?;
                    Ok(Self {
                        name,
                        version,
                        integrity_check: None,
                    })
                }
            }
            None => Err(PackageManagerFieldError::InvalidField {
                field_value: val.to_owned(),
            }),
        }
    }
}

impl Display for PackageManagerField {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if let Some(integrity_check) = self.integrity_check.as_ref() {
            write!(f, "{}@{}+{integrity_check}", self.name, self.version)
        } else {
            write!(f, "{}@{}", self.name, self.version)
        }
    }
}

#[derive(Debug, PartialEq)]
pub(crate) enum PackageManagerFieldError {
    InvalidField {
        field_value: String,
    },
    InvalidPackageManager {
        package_manager: String,
        field_value: String,
    },
    Version {
        field_value: String,
        source: VersionError,
    },
}

#[derive(Debug, PartialEq)]
pub(crate) enum PackageManagerFieldPackageManager {
    Npm,
    Pnpm,
    Yarn,
}

impl FromStr for PackageManagerFieldPackageManager {
    type Err = String;

    fn from_str(val: &str) -> Result<Self, Self::Err> {
        match val {
            "npm" => Ok(PackageManagerFieldPackageManager::Npm),
            "pnpm" => Ok(PackageManagerFieldPackageManager::Pnpm),
            "yarn" => Ok(PackageManagerFieldPackageManager::Yarn),
            _ => Err(val.to_owned()),
        }
    }
}

impl Display for PackageManagerFieldPackageManager {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                PackageManagerFieldPackageManager::Npm => "npm",
                PackageManagerFieldPackageManager::Pnpm => "pnpm",
                PackageManagerFieldPackageManager::Yarn => "yarn",
            }
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::error_handling::test_util::{assert_error_snapshot, create_json_error};
    use serde_json::json;

    #[test]
    fn test_parse_package_manager_field_with_name_and_version() {
        assert_eq!(
            PackageManagerField::from_str("npm@1.2.3"),
            Ok(PackageManagerField {
                name: "npm".parse().unwrap(),
                version: Version::parse("1.2.3").unwrap(),
                integrity_check: None
            })
        );
        assert_eq!(
            PackageManagerField::from_str("pnpm@4.5.6"),
            Ok(PackageManagerField {
                name: "pnpm".parse().unwrap(),
                version: Version::parse("4.5.6").unwrap(),
                integrity_check: None
            })
        );
        assert_eq!(
            PackageManagerField::from_str("yarn@7.8.9"),
            Ok(PackageManagerField {
                name: "yarn".parse().unwrap(),
                version: Version::parse("7.8.9").unwrap(),
                integrity_check: None
            })
        );
    }

    #[test]
    fn test_parse_package_manager_field() {
        assert_eq!(
            PackageManagerField::from_str("pnpm@9.11.0+sha512.0a203ffaed5a3f63242cd064c8fb5892366c103e328079318f78062f24ea8c9d50bc6a47aa3567cabefd824d170e78fa2745ed1f16b132e16436146b7688f19b"),
            Ok(PackageManagerField {
                name: "pnpm".parse().unwrap(),
                version: Version::parse("9.11.0").unwrap(),
                integrity_check: Some("sha512.0a203ffaed5a3f63242cd064c8fb5892366c103e328079318f78062f24ea8c9d50bc6a47aa3567cabefd824d170e78fa2745ed1f16b132e16436146b7688f19b".to_string())
            })
        );
    }

    #[test]
    fn read_valid_package_with_node_engine() {
        let package_json = PackageJson(json!({
            "engines": {
                "node": "16.0.0"
            }
        }));
        assert_eq!(
            package_json.node_engine().unwrap().unwrap().to_string(),
            Requirement::parse("16.0.0").unwrap().to_string()
        );
    }

    #[test]
    fn read_error_message() {
        assert_error_snapshot(&package_json_read_error_message(
            Path::new("./package.json"),
            &std::io::Error::other("test I/O error blah"),
        ));
    }

    #[test]
    fn parse_error_message() {
        assert_error_snapshot(&package_json_parse_error_message(
            Path::new("./package.json"),
            &create_json_error(),
        ));
    }
}
