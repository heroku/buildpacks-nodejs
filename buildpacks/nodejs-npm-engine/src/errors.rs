use crate::install_npm::NpmEngineLayerError;
use crate::BUILDPACK_NAME;
use crate::{node, npm};
use bullet_stream::state::Bullet;
use bullet_stream::{style, Print};
use heroku_nodejs_utils::package_json::PackageJsonError;
use heroku_nodejs_utils::vrs::Requirement;
use indoc::formatdoc;
use std::fmt::Display;
use std::io::{stderr, Stderr};

const USE_DEBUG_INFORMATION_AND_RETRY_BUILD: &str = "\
Use the debug information above to troubleshoot and retry your build.";

const SUBMIT_AN_ISSUE: &str = "\
If the issue persists and you think you found a bug in the buildpack then reproduce the issue \
locally with a minimal example and open an issue in the buildpack's GitHub repository with the details.";

#[derive(Debug)]
pub(crate) enum NpmEngineBuildpackError {
    PackageJson(PackageJsonError),
    MissingNpmEngineRequirement,
    InventoryParse(toml::de::Error),
    NpmVersionResolve(Requirement),
    NpmEngineLayer(NpmEngineLayerError),
    NodeVersion(node::VersionError),
    NpmVersion(npm::VersionError),
}

pub(crate) fn on_error(error: libcnb::Error<NpmEngineBuildpackError>) {
    let logger = Print::new(stderr()).without_header();
    match error {
        libcnb::Error::BuildpackError(buildpack_error) => {
            on_buildpack_error(buildpack_error, logger);
        }
        framework_error => on_framework_error(&framework_error, logger),
    }
}

fn on_buildpack_error(error: NpmEngineBuildpackError, logger: Print<Bullet<Stderr>>) {
    match error {
        NpmEngineBuildpackError::PackageJson(e) => on_package_json_error(e, logger),
        NpmEngineBuildpackError::MissingNpmEngineRequirement => {
            on_missing_npm_engine_requirement_error(logger);
        }
        NpmEngineBuildpackError::InventoryParse(e) => on_inventory_parse_error(&e, logger),
        NpmEngineBuildpackError::NpmVersionResolve(requirement) => {
            on_npm_version_resolve_error(&requirement, logger);
        }
        NpmEngineBuildpackError::NpmEngineLayer(e) => on_npm_engine_layer_error(e, logger),
        NpmEngineBuildpackError::NodeVersion(e) => on_node_version_error(e, logger),
        NpmEngineBuildpackError::NpmVersion(e) => on_npm_version_error(e, logger),
    }
}

fn on_package_json_error(error: PackageJsonError, logger: Print<Bullet<Stderr>>) {
    match error {
        PackageJsonError::AccessError(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    Error reading {package_json}.

                    This buildpack requires {package_json} to complete the build but the file can’t be read. 
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", package_json = style::value("package.json")});
        }
        PackageJsonError::ParseError(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    Error reading {package_json}.

                    This buildpack requires {package_json} to complete the build but the file \
                    can’t be parsed. Ensure {npm_install} runs locally to check the formatting in your file. 
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", package_json = style::value("package.json"), npm_install = style::value("npm install") });
        }
    }
}

fn on_missing_npm_engine_requirement_error(logger: Print<Bullet<Stderr>>) {
    logger.error(formatdoc! {"
        Missing {engines_key} key in {package_json}.
        
        This buildpack requires the `engines.npm` key to determine which engine versions to install.

        Retry your build. 

        {SUBMIT_AN_ISSUE}
    ", engines_key = style::value("engines.npm"), package_json = style::value("package.json") });
}

fn on_inventory_parse_error(error: &toml::de::Error, logger: Print<Bullet<Stderr>>) {
    print_error_details(logger, &error).error(formatdoc! {"
            Failed to load available {npm} versions.

            An unexpected error occurred while loading the available {npm} versions.
        
            {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

            {SUBMIT_AN_ISSUE}
        ", npm = style::value("npm") });
}

fn on_npm_version_resolve_error(requirement: &Requirement, logger: Print<Bullet<Stderr>>) {
    logger.error(formatdoc! {"
            Error resolving requested {npm} version {requested_version}.
            
            Can’t find the `npm` version that matches the requested version declared in {package_json} ({requested_version}).
    
            Verify that the requested version range matches a published version of {npm} by checking \
            https://www.npmjs.com/package/npm?activeTab=versions or trying the following command:
            
                $ npm show 'npm@{requirement}' versions
    
            Update the {engines_key} field in {package_json} to a single version or version range that \
            includes a published {npm} version.
    
            {SUBMIT_AN_ISSUE}
        ",
        npm = style::value("npm"),
        requested_version = style::value(requirement.to_string()),
        package_json = style::value("package.json"),
        engines_key = style::value("engines.npm")
    });
}

fn on_npm_engine_layer_error(error: NpmEngineLayerError, logger: Print<Bullet<Stderr>>) {
    match error {
        NpmEngineLayerError::Download(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    Failed to download {npm}.

                    An unexpected error occurred while downloading the {npm} package. This error can occur due to an unstable network connection.

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::OpenTarball(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    An unexpected error occurred while opening the downloaded {npm} package file. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::DecompressTarball(e) => {
            print_error_details(logger, &e)
                .error(formatdoc! {"
                    An unexpected error occurred while extracting the contents of the downloaded {npm} package file.

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::RemoveExistingNpmInstall(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    An unexpected error occurred while removing the existing {npm} installation. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm") });
        }
        NpmEngineLayerError::InstallNpm(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    An unexpected error occurred while installing the downloaded {npm} package. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
            ", npm = style::value("npm") });
        }
    }
}

fn on_node_version_error(error: node::VersionError, logger: Print<Bullet<Stderr>>) {
    match error {
        node::VersionError::Command(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to determine {node} version information.
    
                    An unexpected error occurred while executing {node_version}. 

                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}

                    {SUBMIT_AN_ISSUE}
                ", node = style::value("Node"), node_version = style::value(e.name()) });
        }
        node::VersionError::Parse(stdout, e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to parse {node} version information.
        
                    An unexpected error occurred while parsing version information from {output}. 
                    
                    {SUBMIT_AN_ISSUE}
                ", node = style::value("Node"), output = style::value(stdout) });
        }
    }
}

fn on_npm_version_error(error: npm::VersionError, logger: Print<Bullet<Stderr>>) {
    match error {
        npm::VersionError::Command(e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to determine {npm} version information.

                    An unexpected error occurred while executing {npm_version}.  
                    
                    {USE_DEBUG_INFORMATION_AND_RETRY_BUILD}
        
                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm"), npm_version = style::value(e.name())});
        }
        npm::VersionError::Parse(stdout, e) => {
            print_error_details(logger, &e).error(formatdoc! {"
                    Failed to parse {npm} version information.
        
                    An unexpected error occurred while parsing version information from {output}. 
                    
                    {SUBMIT_AN_ISSUE}
                ", npm = style::value("npm"), output = style::value(stdout) });
        }
    }
}

fn on_framework_error(
    error: &libcnb::Error<NpmEngineBuildpackError>,
    logger: Print<Bullet<Stderr>>,
) {
    print_error_details(logger, &error)
        .error(formatdoc! {"
            {buildpack_name} internal error.
    
            The framework used by this buildpack encountered an unexpected error.

            If you can't deploy to Heroku due to this issue, check the official Heroku Status page at \
            status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.

            If the issue persists and you think you found a bug in the buildpack or framework, reproduce \
            the issue locally with a minimal example. Open an issue in the buildpack's GitHub repository \
            and include the details.

        ", buildpack_name = style::value(BUILDPACK_NAME) });
}

fn print_error_details(
    logger: Print<Bullet<Stderr>>,
    error: &impl Display,
) -> Print<Bullet<Stderr>> {
    logger
        .bullet(style::important("DEBUG INFO:"))
        .sub_bullet(error.to_string())
        .done()
}
