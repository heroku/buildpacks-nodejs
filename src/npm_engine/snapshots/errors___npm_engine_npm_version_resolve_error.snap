---
source: buildpacks/nodejs-npm-engine/src/errors.rs
---
! Error resolving requested `npm` version `1.2.3`
!
! The requested npm version could not be resolved to a known release in this buildpack's inventory of npm releases.
!
! Suggestions:
! - Confirm if this is a valid npm release at https://www.npmjs.com/package/npm?activeTab=versions or by running `npm show 'npm@1.2.3' versions`
! - Check if this buildpack includes the requested npm version in its inventory file at https://github.com/heroku/buildpacks-nodejs/blob/main/buildpacks/nodejs-npm-engine/inventory.toml
! - Update the `engines.npm` field in `package.json` to a single version or version range that includes a published `npm` version.
!
! If the issue persists and you think you found a bug in the buildpack, reproduce the issue locally with a minimal example. Open an issue in the buildpack's GitHub repository and include the details here:
! https://github.com/heroku/buildpacks-nodejs/issues
